### PR TITLE
Detect all introduced names

### PR DESCRIPTION
**Pull request by LauraSManns**

### CODE REVIEW

File: xun/functions/util.py [hunk 1]
```diff
@@ -284,10 +285,53 @@ def stmt_introduced_names(stmt):
...
+        if isinstance(stmt, ast.ClassDef):
+            return frozenset({stmt.name})
+        if isinstance(stmt, ast.AnnAssign):
+            return frozenset({stmt.target.id})
+        if isinstance(stmt, ast.Delete):
+            return frozenset()
+        if isinstance(stmt, (ast.For, ast.AsyncFor, ast.While, ast.If)):
+            a = set()
+            if isinstance(stmt, (ast.For, ast.AsyncFor)):
+                a = {stmt.target.id}
```
Comment (JensGM) [inline xun/functions/util.py:301]: I think for statements can have multiple targets

File: xun/functions/util.py [hunk 2]
```diff
@@ -275,19 +275,70 @@ def shape_to_ast_tuple(shape):
...
+        targets = stmt.targets
+        return assignment_target_introduced_names(targets)
+    if isinstance(stmt, ast.AugAssign):
+        return frozenset()
+    if isinstance(stmt, ast.AnnAssign):
+        return frozenset({stmt.target.id})
+    if isinstance(stmt, (ast.For, ast.AsyncFor)):
+        names = {stmt.target.id}
+        add_names_from_body(names, stmt.body)
+        return frozenset(names)
```
Comment (mrSundvoll) [inline xun/functions/util.py:304]: We need to consider the case where the `stmt.target` is a tuple, e.g. `for i,j in d.items():`.

```suggestion
    def names_in_target(target):
        names = set()
        for elt in target.elts:
            if isinstance(elt, ast.Name):
                names.add(elt.id)
            elif isinstance(elt. ast.Tuple):
                names.update(names_in_target(elt)
            else:
                raise TypeError(f'Invalid AST type {type(elt)}')
        return names
        
    if isinstance(stmt, (ast.For, ast.AsyncFor)):
        names = names_in_target(stmt.target)
        add_names_from_body(names, stmt.body)
        return frozenset(names)
```

**Review by mrSundvoll — CHANGES_REQUESTED** [review]

**Review by mrSundvoll — APPROVED** [review]

File: xun/functions/util.py [hunk 3]
```diff
@@ -275,19 +275,85 @@ def shape_to_ast_tuple(shape):
...
+    if isinstance(stmt, ast.Raise):
+        return frozenset()
+    if isinstance(stmt, ast.Try):
+        names = set()
+        add_names_from_body(names, stmt.body)
+        for handler in stmt.handlers:
+            add_names_from_body(names, handler.body)
+        add_names_from_body(names, stmt.orelse)
+        add_names_from_body(names, stmt.finalbody)
+        return frozenset(names)
```
Comment (JensGM) [inline xun/functions/util.py:340]:
```suggestion
    if isinstance(stmt, ast.Try):
        names = {stmt_introduced_names(s) for s in stmt.body}
        names |= {stmt_introduced_names(s) for s in handler for handler in stmt.handlers}
        names |= {stmt_introduced_names(s) for s in stmt.orelse}
        names |= {stmt_introduced_names(s) for s in stmt.finalbody}
        return frozenset(names)
```

```suggestion
    if isinstance(stmt, ast.Try):
        return frozenset().union(
            {stmt_introduced_names(s) for s in stmt.body},
            {stmt_introduced_names(s) for s in h for h in stmt.handlers},
            {stmt_introduced_names(s) for s in stmt.orelse},
            {stmt_introduced_names(s) for s in stmt.finalbody},
        )
```

File: xun/functions/util.py [hunk 4]
```diff
@@ -275,19 +275,85 @@ def shape_to_ast_tuple(shape):
...
+    if isinstance(stmt, ast.AnnAssign):
+        return frozenset({stmt.target.id})
+    if isinstance(stmt, (ast.For, ast.AsyncFor)):
+        names = names_in_target(stmt.target)
+        add_names_from_body(names, stmt.body)
+        return frozenset(names)
+    if isinstance(stmt, (ast.While, ast.If)):
+        names = set()
+        add_names_from_body(names, stmt.body)
+        return frozenset(names)
```
Comment (JensGM) [inline xun/functions/util.py:320]:
```suggestion
    if isinstance(stmt, (ast.While, ast.If)):
        return frozenset(stmt_introduced_names(s) for s in stmt.body)
```

File: xun/functions/util.py [hunk 5]
```diff
@@ -275,19 +275,85 @@ def shape_to_ast_tuple(shape):
...
+    if isinstance(stmt, ast.Assign):
+        return assignment_target_introduced_names(stmt.targets)
+    if isinstance(stmt, ast.AugAssign):
+        return frozenset()
+    if isinstance(stmt, ast.AnnAssign):
+        return frozenset({stmt.target.id})
+    if isinstance(stmt, (ast.For, ast.AsyncFor)):
+        names = names_in_target(stmt.target)
+        add_names_from_body(names, stmt.body)
+        return frozenset(names)
```
Comment (JensGM) [inline xun/functions/util.py:316]:
```suggestion
        names = frozenset()
        names |= assignment_target_introduced_names(stmt.target)
        names |= {statement_introduced_names(s) for s in stmt.body}
        return names
```

File: xun/functions/util.py [hunk 6]
```diff
@@ -275,19 +275,85 @@ def shape_to_ast_tuple(shape):
...
+    if isinstance(stmt, (ast.With, ast.AsyncWith)):
+        names = set()
+        for item in stmt.items:
+            if isinstance(item.optional_vars, ast.Name):
+                names.add(item.optional_vars.id)
+            if isinstance(item.optional_vars, ast.Tuple):
+                for el in item.optional_vars.elts:
+                    names.add(el.id)
+        add_names_from_body(names, stmt.body)
+        return frozenset(names)
```
Comment (JensGM) [inline xun/functions/util.py:330]:
```suggestion
    if isinstance(stmt, (ast.With, ast.AsyncWith)):
        names = frozenset()
        names |= {assignment_target_introduced_names(item.optional_vars)
                  for item in stmt.items if item.optional_vars is not None}
        names |= {stmt_introduced_names(s) for s in stmt.body}
        return names
```

File: xun/functions/util.py [hunk 7]
```diff
@@ -275,19 +275,85 @@ def shape_to_ast_tuple(shape):
...
+    if isinstance(stmt, ast.Return):
+        return frozenset()
+    if isinstance(stmt, ast.Delete):
+        return frozenset()
+    if isinstance(stmt, ast.Assign):
+        return assignment_target_introduced_names(stmt.targets)
+    if isinstance(stmt, ast.AugAssign):
+        return frozenset()
+    if isinstance(stmt, ast.AnnAssign):
+        return frozenset({stmt.target.id})
```
Comment (JensGM) [inline xun/functions/util.py:312]:
```suggestion
        return assignment_target_introduced_names(stmt.target)
```

File: xun/functions/util.py [hunk 8]
```diff
@@ -275,19 +275,85 @@ def shape_to_ast_tuple(shape):
...
+            return {target.id}
+        names = set()
+        for elt in target.elts:
+            if isinstance(elt, ast.Name):
+                names.add(elt.id)
+            elif isinstance(elt. ast.Tuple):
+                names.update(names_in_target(elt))
+            else:
+                raise TypeError(f'Invalid AST type {type(elt)}')
+        return names
```
Comment (JensGM) [inline xun/functions/util.py:297]:
```suggestion
```

**Review by JensGM — CHANGES_REQUESTED** [review]

File: xun/functions/util.py [hunk 9]
```diff
@@ -275,19 +275,61 @@ def shape_to_ast_tuple(shape):
...
+        names |= assignment_target_introduced_names(stmt.target)
+        names |= {statement_introduced_names(s) for s in stmt.body}
+        return names
+    if isinstance(stmt, (ast.While, ast.If)):
+        return frozenset(stmt_introduced_names(s) for s in stmt.body)
+    if isinstance(stmt, (ast.With, ast.AsyncWith)):
+        names = frozenset()
+        names |= {assignment_target_introduced_names(item.optional_vars)
+                  for item in stmt.items if item.optional_vars is not None}
+        names |= {stmt_introduced_names(s) for s in stmt.body}
```
Comment (JensGM) [inline xun/functions/util.py:307]:
```suggestion
        names |= chain(stmt_introduced_names(s) for s in stmt.body)
```

**Review by JensGM — APPROVED** [review]
When tests run, if you rebase I'm happy!

**Review by mrSundvoll — APPROVED** [review]

**Pull request merged**